File: Game2D/Engine/Laser.cs
using System.Windows.Controls;
using System.Windows.Media.Imaging;
using System;

namespace Game2D.Engine
{
    public class Laser : GameObject
    {
        private double dx, dy;
        private const double Speed = 8;
        private int damage = 75;
        public Laser(Canvas canvas, double x, double y, double angle)
        {
            X = x;
            Y = y;
            dx = Math.Cos(angle) * Speed;
            dy = Math.Sin(angle) * Speed;
            Sprite = new Image
            {
                Source = new BitmapImage(new Uri("pack://application:,,,/Project/images/laser.png")),
                Width = 48,
                Height = 48
            };
            Sprite.RenderTransform = new System.Windows.Media.RotateTransform(angle * 180 / Math.PI, 24, 24);
        }
        public override void Update()
        {
            X += dx;
            Y += dy;
            var mainWindow = System.Windows.Application.Current.MainWindow as Game2D.MainWindow;
            if (mainWindow != null && mainWindow.GameCanvas != null)
            {
                double w = mainWindow.GameCanvas.ActualWidth;
                double h = mainWindow.GameCanvas.ActualHeight;
                if (X < -Sprite.Width || X > w || Y < -Sprite.Height || Y > h)
                {
                    if (Sprite.Parent is Canvas canvas) canvas.Children.Remove(Sprite);
                    IsActive = false;
                }
            }
            var world = Game2D.MainWindow.CurrentGameWorld;
            if (world != null)
            {
                foreach (var obj in world.Objects)
                {
                    if (obj is EnemyBeast enemy && enemy.IsActive && this.GetBounds().IntersectsWith(enemy.GetBounds()))
                    {
                        enemy.TakeDamage(damage);
                        if (Sprite.Parent is Canvas canvas) canvas.Children.Remove(Sprite);
                        this.IsActive = false;
                        break;
                    }
                    if (obj is CosmoEnemy cosmo && cosmo.IsActive && this.GetBounds().IntersectsWith(cosmo.GetBounds()))
                    {
                        cosmo.TakeDamage(damage);
                        if (Sprite.Parent is Canvas canvas) canvas.Children.Remove(Sprite);
                        this.IsActive = false;
                        break;
                    }
                    if (obj is ShotgunEnemy shot && shot.IsActive && this.GetBounds().IntersectsWith(shot.GetBounds()))
                    {
                        shot.TakeDamage(damage);
                        if (Sprite.Parent is Canvas canvas) canvas.Children.Remove(Sprite);
                        this.IsActive = false;
                        break;
                    }
                    if (obj is RocketEnemy rocket && rocket.IsActive && this.GetBounds().IntersectsWith(rocket.GetBounds()))
                    {
                        rocket.TakeDamage(damage);
                        if (Sprite.Parent is Canvas canvas) canvas.Children.Remove(Sprite);
                        this.IsActive = false;
                        break;
                    }
                    if (obj is Zombi zombi && zombi.IsActive && this.GetBounds().IntersectsWith(zombi.GetBounds()))
                    {
                        zombi.TakeDamage(damage);
                        if (Sprite.Parent is Canvas canvas) canvas.Children.Remove(Sprite);
                        this.IsActive = false;
                        break;
                    }
                    if (obj is Boss boss && this.GetBounds().IntersectsWith(boss.GetBounds()))
                    {
                        boss.TakeDamage(damage);
                        if (Sprite.Parent is Canvas canvas) canvas.Children.Remove(Sprite);
                        this.IsActive = false;
                        break;
                    }
                    if (obj is Boss_Hand1 hand1 && this.GetBounds().IntersectsWith(hand1.GetBounds()))
                    {
                        hand1.TakeDamage(damage);
                        if (Sprite.Parent is Canvas canvas) canvas.Children.Remove(Sprite);
                        this.IsActive = false;
                        break;
                    }
                    if (obj is Boss_Hand2 hand2 && this.GetBounds().IntersectsWith(hand2.GetBounds()))
                    {
                        hand2.TakeDamage(damage);
                        if (Sprite.Parent is Canvas canvas) canvas.Children.Remove(Sprite);
                        this.IsActive = false;
                        break;
                    }
                    if (obj is Wall || obj is Wall_gorizont || obj is Wall_vertical)
                    {
                        if (this.GetBounds().IntersectsWith(obj.GetBounds()))
                        {
                            if (Sprite.Parent is Canvas canvas) canvas.Children.Remove(Sprite);
                            this.IsActive = false;
                            break;
                        }
                    }
                }
            }
        }
    }
} 

File: Game2D/Project/Laser.java
import greenfoot.*;  // (World, Actor, GreenfootImage, Greenfoot and MouseInfo)

public class Laser extends Actor
{
    private double dx,dy;
    private int speed=8;
    
    public Laser(double angle){
        setRotation((int)Math.toDegrees(angle)+90);
        dx=Math.cos(angle)*speed;
        dy=Math.sin(angle)*speed;
        
        setImage("laser.png");
    }
    public void act()
    {
        if (isAtEdge() || checkCollision()){
            World world=getWorld();
            if (world!=null) world.removeObject(this);
            return;
        }
        
        checkCollision();
        if (!isInWorld()) return;
        MoveLaser();
        
    }
    
    private boolean isInWorld(){
        return getWorld() != null;
    }
    
    private void MoveLaser(){
        setLocation((int)(getX()+dx),(int)(getY()+dy));
    }
    
    private boolean checkCollision() {
        if (getWorld() == null) return false;

        // Столкновение со стенами
        if (getOneIntersectingObject(Wall_gorizont.class) != null ||
            getOneIntersectingObject(Wall_vertical.class) != null ||
            getOneIntersectingObject(Cube.class) != null) {
            return true;
        }

        // Столкновение с зомби
        zombi zombie = (zombi) getOneIntersectingObject(zombi.class);
        if (zombie != null) {
            zombie.takeDamage(100); // ← Используем метод убийства
            return true;
        }

        // Столкновение с обычным врагом
        Enemy_beast enemy = (Enemy_beast) getOneIntersectingObject(Enemy_beast.class);
        if (enemy != null) {
            MyWorld world = (MyWorld) getWorld();
            if (world != null) {
                world.incrementKillCount();
                world.removeObject(enemy);
            }
            return true;
        }
        
        RocketEnemy rocketEnemy = (RocketEnemy) getOneIntersectingObject(RocketEnemy.class);
        if (rocketEnemy != null) {
            rocketEnemy.takeDamage(75); 
            MyWorld world = (MyWorld) getWorld();
            if (world != null) {
                world.incrementKillCount(); 
            }
            return true;
        }
        
        CosmoEnemy cosmo = (CosmoEnemy) getOneIntersectingObject(CosmoEnemy.class);
        if (cosmo != null) {
            cosmo.takeDamage(75); 
            MyWorld world = (MyWorld) getWorld();
            if (world != null) {
                world.incrementKillCount(); 
            }
            return true;
        }
        
        ShotgunEnemy shot = (ShotgunEnemy) getOneIntersectingObject(ShotgunEnemy.class);
        if (shot != null) {
            shot.takeDamage(75); 
            MyWorld world = (MyWorld) getWorld();
            if (world != null) {
                world.incrementKillCount(); 
            }
            return true;
        }
        
        Boss boss = (Boss) getOneIntersectingObject(Boss.class);
          if (boss != null) {
          boss.takeDamage(18); // Наносим урон боссу (можно настроить значение)
          return true; // Пуля исчезает после попадания
        }

        return false;
    }
}
